pl_bolts/models/detection/yolo/yolo_layers.py

```diff
@@ -0,0 +1,553 @@
+from typing import Callable, Dict, List, Optional, Tuple
+
+import torch
+from pytorch_lightning.utilities.exceptions import MisconfigurationException
+from torch import nn, Tensor
+
+from pl_bolts.utils.warnings import warn_missing_pkg
+
+try:
+    from torchvision.ops import box_area, box_iou
+except ModuleNotFoundError:
+    warn_missing_pkg('torchvision')  # pragma: no-cover
+    _TORCHVISION_AVAILABLE = False
+else:
+    _TORCHVISION_AVAILABLE = True
+
+
+def _corner_coordinates(xy, wh):
+    """
+    Converts box center points and sizes to corner coordinates.
+
+    Args:
+        xy (Tensor): Center coordinates. Tensor of size `[..., 2]`.
+        wh (Tensor): Width and height. Tensor of size `[..., 2]`.
+
+    Returns:
+        boxes (Tensor): A matrix of (x1, y1, x2, y2) coordinates.
+    """
+    half_wh = wh / 2
+    top_left = xy - half_wh
+    bottom_right = xy + half_wh
+    return torch.cat((top_left, bottom_right), -1)
+
+
+def _aligned_iou(dims1, dims2):
+    """
+    Calculates a matrix of intersections over union from box dimensions, assuming that the boxes
+    are located at the same coordinates.
+
+    Args:
+        dims1 (Tensor[N, 2]): width and height of N boxes
+        dims2 (Tensor[M, 2]): width and height of M boxes
+
+    Returns:
+        iou (Tensor[N, M]): the NxM matrix containing the pairwise IoU values for every element in
+            `dims1` and `dims2`
+    """
+    area1 = dims1[:, 0] * dims1[:, 1]  # [N]
+    area2 = dims2[:, 0] * dims2[:, 1]  # [M]
+
+    inter_wh = torch.min(dims1[:, None, :], dims2)  # [N, M, 2]
+    inter = inter_wh[:, :, 0] * inter_wh[:, :, 1]  # [N, M]
+    union = area1[:, None] + area2 - inter  # [N, M]
+
+    return inter / union
+
+
+def _elementwise_iou(boxes1: Tensor, boxes2: Tensor) -> Tensor:
```

Comment:
hmm, this is same as `torchvision.ops.box_iou` ? I guess ?

Reply:
@oke-aditya It's basically the same thing, but performs the operation elementwise on two sets of N boxes, producing N results. The torchvision op performs the operation on every pair formed from two sets of boxes, producing an N*M matrix. In other words, this is the same as taking the diagonal of `torchvision.ops.box_iou`, however, that would take quadratic time. The most generic and consistent way would probably be that the torchvision `box_iou` and `generalized_box_iou` functions would perform the elementwise operation, and for creating a pairwise matrix one would add empty dimensions to the inputs (or include an optional argument in those functions to perform the elementwise operation).

Reply:
So, here we have two options (right?):
1. to use `_elementwise_iou` as is, or
2. to remove `_elementwise_iou` and use `torchvision.ops.box_iou` (but increase computational cost from `O(N)` to `O(N^2)`)

@oke-aditya @PyTorchLightning/bolts-contributors @Borda Any thoughts on this? (I'm neutral on this.)

Reply:
It is fine to increase a slight computational cost. Since these are tensor operations, they are fast and batched, it will not scale up so soon.

Reply:
I might add a third option, suggesting a change in the torchvision interface that supports both variants. But it's better to merge this PR first and improve it in future PRs, so I'll do like @oke-aditya suggests and use the current torchvision interface for now. At least then we'll know that it works correctly.

Reply:
I'm using torchvision now. By the way, the computational cost is quadratic in the _number of targets per image_, which is typically not very large. `generalized_box_iou` is not yet supported by the version of torchvision used by the PL Bolts test suite, so I added a check.

Reply:
> number of targets per image, which is typically not very large.

Sounds reasonable! Thanks for the update!